docs/tutorials/Containers.md

```diff
@@ -0,0 +1,475 @@
+## Containers
+
+Containers are self-contained software archives, that hold everything from the tool you want to run, to libraries and runtimes it requires, and the operating system it runs on.
+ 
+Best Practise WDL and CWL define containers for their tasks to run in, to ensure reproducibility and portability - that running the same task on a different system will run the exact same software. 
+
+Docker images are the most common container format, but it is not advisable for certain systems to run Docker itself, and for this reason Cromwell can be configured to support a number of alternatives.
+
+* [Prerequisites](#prerequisites)
+* [Goals](#goals)
+* [Specifying Containers in your Workflow](#specifying-containers-in-your-workflow)
+* [Docker](#docker)
+    * [Docker on a Local Backend](#docker-on-a-local-backend)
+    * [Docker on Cloud](#docker-on-cloud)
+    * [Docker on HPC](#docker-on-hpc)
+* [Singularity](#singularity)
+    * [Installation](#installation)
+    * [Configuring Cromwell for Singularity](#configuring-cromwell-for-singularity)
+        * [Local environments](#local-environments)
+        * [Job schedulers](#job-schedulers)
+    * [Without Setuid](#without-setuid)
+    * [Singularity Cache](#singularity-cache)
+* [udocker](#udocker)
+    * [Installation](#installation-1)
+    * [Configuration](#configuration)
+    * [Caching](#caching)
+* [Configuration in Detail](#configuration-in-detail)
+    * [Enforcing container requirements](#enforcing-container-requirements)
+    * [Docker Digests](#docker-digests)
+    * [Docker Root](#docker-root)
+    * [Docker Config Block](#docker-config-block)
+* [Best Practices](#best-practices)
+    * [Image Versions](#image-versions)
+* [Notes](#notes)
+    * [How does Cromwell know when a job or container has completed?](#how-does-cromwell-know-when-a-job-or-container-has-completed)
+    * [Cromwell: Run-in-background](#cromwell-run-in-background)
+* [Next Steps](#next-steps)
+
+### Prerequisites
+This tutorial page relies on completing the previous tutorials:
+
+* [Five Minute Introduction](FiveMinuteIntro.md)
+* [Configuration Files](ConfigurationFiles.md)
+* Recommended: [Getting started on HPC clusters](HPCIntro.md)
+
+### Goals
+
+At the end of this tutorial, you'll become familiar with container technologies and how to configure Cromwell to use these independently, or with job schedulers.
+
+
+### Specifying Containers in your Workflow
+
+Containers are specified on a per-task level, this can be achieved in WDL by specifying a [`docker`](https://github.com/openwdl/wdl/blob/master/versions/1.0/SPEC.md#docker) tag in the `runtime` section. For example, specifying that the following WDL script should use the container `ubuntu:latest` can be achieved by:
```

Comment:
> level, *and* this can be achieved

Reply:
I missed replying to this comment, but new text is:
> For example, the following script should run in the `ubuntu:latest` container: